hivemq-edge/src/frontend/src/modules/Workspace/components/nodes/ContextualToolbar.tsx

```diff
@@ -0,0 +1,139 @@
+import { type FC, MouseEventHandler, useMemo } from 'react'
+import { NodeToolbar, type NodeProps, type NodeToolbarProps, Position, Node, Edge, MarkerType } from 'reactflow'
+import { useTranslation } from 'react-i18next'
+import { useTheme } from '@chakra-ui/react'
+import { LuPanelRightOpen } from 'react-icons/lu'
+import { ImMakeGroup } from 'react-icons/im'
+
+import { Adapter, Status } from '@/api/__generated__'
+import { EdgeTypes, Group, IdStubs, NodeTypes } from '@/modules/Workspace/types.ts'
+import IconButton from '@/components/Chakra/IconButton.tsx'
+import useWorkspaceStore from '@/modules/Workspace/hooks/useWorkspaceStore.ts'
+import { getGroupLayout } from '@/modules/Workspace/utils/group.utils.ts'
+import { getThemeForStatus } from '@/modules/Workspace/utils/status-utils.ts'
+import WorkspaceButtonGroup from '@/modules/Workspace/components/parts/WorkspaceButtonGroup.tsx'
+
+type SelectedNodeProps = Pick<NodeProps, 'id'> & Pick<NodeToolbarProps, 'position'>
+interface ContextualToolbarProps extends SelectedNodeProps {
+  onOpenPanel?: MouseEventHandler | undefined
+  children?: React.ReactNode
+}
+
+const ContextualToolbar: FC<ContextualToolbarProps> = ({ id, onOpenPanel, children }) => {
+  const { t } = useTranslation()
+  const { onInsertGroupNode, nodes } = useWorkspaceStore()
+  const theme = useTheme()
+
+  const selectedNodes = nodes.filter((node) => node.selected)
+  const selectedGroupCandidates = useMemo(() => {
+    // TODO[NVL] Should the grouping only be available if ALL nodes match the filter ?
+    const adapters = selectedNodes.filter((e) => e.type === NodeTypes.ADAPTER_NODE && !e.parentId && !e.parentNode)
+    return adapters.length >= 2 ? adapters : undefined
+  }, [selectedNodes])
+
+  const onCreateGroup = () => {
+    if (!selectedGroupCandidates) return
+
+    const groupId = `${IdStubs.GROUP_NODE}@${selectedGroupCandidates.map((e) => e.data.id).join('+')}`
+    const groupTitle = t('workspace.grouping.untitled')
+    const rect = getGroupLayout(selectedGroupCandidates)
+    const newGroupNode: Node<Group, NodeTypes.CLUSTER_NODE> = {
+      id: groupId,
+      type: NodeTypes.CLUSTER_NODE,
+      data: {
+        childrenNodeIds: selectedGroupCandidates.map((e) => e.id),
+        title: groupTitle,
+        isOpen: true,
+        colorScheme: 'red',
+      },
+      style: {
+        width: rect.width,
+        height: rect.height,
+      },
+      position: { x: rect.x, y: rect.y },
+    }
+
+    const groupStatus: Status = {
+      runtime: selectedGroupCandidates.every((e: Node<Adapter>) => e.data.status?.runtime === Status.runtime.STARTED)
+        ? Status.runtime.STARTED
+        : Status.runtime.STOPPED,
+      connection: selectedGroupCandidates.every(
+        (e: Node<Adapter>) => e.data.status?.connection !== Status.connection.CONNECTED
+      )
+        ? Status.connection.CONNECTED
+        : Status.connection.DISCONNECTED,
```

Comment:
am I misunderstanding this? I'm reading:
`if every connection is NOT "CONNECTED", then set connection = CONNECTED, otherwise set it to DISCONNECTED`.

It seems backwards to me 🤔 
If it's on purpose, then maybe add a comment to explain it

Reply:
No, that's a bug. Thanks for spotting that one